lib/Sema/SemaBounds.cpp

```diff
@@ -2145,11 +2153,143 @@ namespace {
 
 
   public:
-    CheckBoundsDeclarations(Sema &S, BoundsExpr *ReturnBounds) : S(S),
+    CheckBoundsDeclarations(Sema &S, Stmt *Body, CFG *Cfg, BoundsExpr *ReturnBounds) : S(S),
       DumpBounds(S.getLangOpts().DumpInferredBounds),
       PointerWidth(S.Context.getTargetInfo().getPointerWidth(0)),
+      Body(Body),
+      Cfg(Cfg),
       ReturnBounds(ReturnBounds) {}
 
+    typedef llvm::SmallPtrSet<const Stmt *, 16> StmtSet;
+
+    void IdentifyChecked(Stmt *S, StmtSet &CheckedStmts, bool InCheckedScope) {
+      if (!S)
+        return;
+
+      if (InCheckedScope)
+        if (isa<Expr>(S) || isa<DeclStmt>(S) || isa<ReturnStmt>(S))
+          CheckedStmts.insert(S);
+
+      if (CompoundStmt *CS = dyn_cast<CompoundStmt>(S))
+        InCheckedScope = CS->isChecked();
+
+      auto Begin = S->child_begin(), End = S->child_end();
+      for (auto I = Begin; I != End; ++I)
+        IdentifyChecked(*I, CheckedStmts, InCheckedScope);
+   }
+
+    // Add any subexpressions of S that occur in TopLevelElems to NestedExprs.
+    void MarkNested(const Stmt *S, StmtSet &NestedExprs, StmtSet &TopLevelElems) {
+      auto Begin = S->child_begin(), End = S->child_end();
+      for (auto I = Begin; I != End; ++I) {
+        const Stmt *Child = *I;
+        if (!Child)
+          continue;
+        if (TopLevelElems.find(Child) != TopLevelElems.end())
+          NestedExprs.insert(Child);
+        MarkNested(Child, NestedExprs, TopLevelElems);
+      }
+   }
+
+  // Identify CFG elements that are statements that are substatements of other
+  // CFG elements.  (CFG elements are the components of basic blocks).  When a
+  // CFG is constructed, subexpressions of top-level expressions may be placed
+  // in separate CFG elements.  This is done for subexpressions of expressions
+  // with control-flow, for example. When checking bounds declarations, we want
+  // to process a subexpression with its enclosing expression. We want to
+  // ignore CFG elements that are substatements of other CFG elements.
```

Comment:
I suggest putting a simple example here to help users understand, like:
Some complex conditional expressions may bring extra control-flow...
x = (x == 0) ? f1() : f2();